File: 128.longest-consecutive-sequence.ts
// @leet start
function longestConsecutive(nums: number[]): number {
	let max_length: number = 0;
	let memo = {};

	for (let i = 0; i < nums.length; i++) {
		let val = nums[i];
		if (!(val in memo)) {
			memo[val] = 1;
		}
	}

	for (const key in memo) {
		let key_val = Number(key);
		if ((key_val - 1) in memo) {
			memo[key] += memo[key_val];
		}
	}

	for (const key in memo) {
		max_length = Math.max(memo[key], max_length);
	}

	return max_length;
    
};
// @leet end

File: 128.longest-consecutive-sequence.cpp
// @leet start
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        int result = 0;
        unordered_map<int, int> map;
        for (int num: nums) {
            map[num] = 1;
        }

        for (int num: nums) {
            result = max(result, longestHelper(num, map));
        }


        return result;
    }

    int longestHelper(int num, unordered_map<int,int> &map) {
        if (map.find(num) == map.end()) {
            return 0;
        }

        if (map[num] == 1) {
            map[num] = longestHelper(num + 1, map) + 1;
        }

        return map[num];
    }
};
// @leet end
